### PR TITLE
Unified PR for WiFi hotspot testing of channel-changing, country code, etc

### PR DESCRIPTION
**Pull request by holta**
@EMG70 is helping a lot with testing and this new unified PR helps him install both of these more efficiently:

- PR #3542 
- PR #3546 

Huge thanks to @EMG70 for the diligent hard work!

Here are the new test install instructions for any new OS's / new machines:

```
curl iiab.io/install.txt | bash -s 3549
```

### CODE REVIEW

**Comment by jvonau** [comment]
adm-run-roles-tmp.yml ??

File: adm-run-roles-tmp.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+---
```
Comment (jvonau) [inline adm-run-roles-tmp.yml]: This should not be present

**Comment by jvonau** [comment]
Um I closed the other PR for a reason think c49adcf39596b7b39f673c24db605fd9837e3537 went too far and broke netplan-systemd, I'd revert that one.

**Comment by holta** [comment]
> think [c49adcf](https://github.com/iiab/iiab/commit/c49adcf39596b7b39f673c24db605fd9837e3537) went too far and broke netplan-systemd, I'd revert that one.

Thanks.  Is [fb2a9c8](https://github.com/holta/iiab/commit/fb2a9c82222f39c3d1a839454b335e71a1ee1ea1) plausible, or close enough for initial testing?

**Comment by holta** [comment]
Oops, earlier I did not see:

- PR holta/iiab#511

( Above mini-PR 511 is now merged into this PR 3549, overwriting [fb2a9c8](https://github.com/holta/iiab/commit/fb2a9c82222f39c3d1a839454b335e71a1ee1ea1) edits on the same file roles/network/templates/hostapd/iiab-wifi-test.service.j2 )

**Comment by holta** [comment]
Not the final word — but good news @EMG70 made progress testing _(an earlier version, from about 8h ago)_ of the changes introduced by this PR — that he retrofitted[*] onto his IIAB / Debian 11 laptop (with graphical desktop).  Quick results:

- Changing the hotspot's WiFi channel from 1 to 6 to 11 worked for him, without any need to reboot (so far so good!)
- In coming days if he has time, he'll try a fresh install of this PR (#3549) on 64-bit Raspberry Pi OS Lite on RPi 400 — to confirm things more throughly, e.g.:
  - Does WiFi Channel-changing continues to well well?
  - Do `iiab-hotspot-off` and `iiab-hotspot-on` work well?
  - Does `iiab-test-wifi` work well?

Apologies his [iiab-diagnostics](https://github.com/iiab/iiab/blob/master/scripts/iiab-diagnostics.README.md) don't show all the details as he's juggling many responsibilities, but here's just a rough summary for context: http://sprunge.us/BIh3wH?en

[*] I didn't even realize retrofitting PR's (pull requests) onto an existing IIAB can (sometimes!) be so easy, but FYI he succeeded in applying PR's onto his existing IIAB (Dell laptop where he installed IIAB on Debian 11 about 4 weeks ago).  Long story short, applying PR's to an existing IIAB is (sometimes!) as easy as what he did here:

1) Delete `/etc/iiab/install-flags/iiab-complete`
2) Run `sudo iiab 3542 3546`
   1) Going forward, he would instead use `sudo iiab 3549` to install this unified PR (3549) — if in the end there's once again a need — e.g. to install this test functionality on top of a pre-existing IIAB.

**Comment by jvonau** [comment]
FWIW I usually do a `git pull` `iiab-network` without any real issues at all.
> Apr 24 22:01:48 box wpa_supplicant[672]: wlp2s0: CTRL-EVENT-SCAN-FAILED ret=-95

Think these entries occur because of using 'wifi_up_down: False' with a desktop installed. The pick list that would be shown within the wifi applet is populated by scanning the wifi spectrum but that is not possible when the only wifi interface is assigned to hostapd. Think you will find these entries stop when 'wifi_up_down: True' is used as the desktop would have an interface to use for the scanning.

**Comment by jvonau** [comment]
https://github.com/holta/iiab/pull/512

**Comment by holta** [comment]
@EMG70 tested a fresh install of IIAB with this PR (#3549) on Ubuntu Desktop 23.04 on Raspberry Pi 400.

In this case with: `wifi_up_down: True`

He was surprised to discover that its WiFi hotspot channel could be changed **without** running iiab-hotspot-off, as follows:

- Edit `host_channel:` in `/etc/iiab/local_vars.yml` changing it from 6 to 1 (or 11)
- Run `sudo iiab-network`
- Check to see if new `channel` appears in `/etc/hostapd/hostapd.conf`

(He realizes this is not the official method, but tested it repeatedly, and in this case it does work!)

iiab-diagnostics: http://sprunge.us/r0BpIy?en

FWIW the output of `journalctl -b 0 -u networkd-dispatcher` below Line 1408 looks potentially abnormal ?  (Or maybe this is normal!)

**Comment by jvonau** [comment]
> @EMG70 tested a fresh install of IIAB with this PR (#3549) on Ubuntu Desktop 23.04 on Raspberry Pi 400.
> 
> In this case with: `wifi_up_down: True`
> 
> He was surprised to discover that its WiFi hotspot channel could be changed **without** running iiab-hotspot-off, as follows:
>

Should work fine without any upstream wifi, there is no channel contention for the RF radio. 
>     * Edit `host_channel:` in `/etc/iiab/local_vars.yml` changing it from 6 to 1 (or 11)
> 
>     * Run `sudo iiab-network`
> 
>     * Check to see if new `channel` appears in `/etc/hostapd/hostapd.conf`
> 
> 
> (He realizes this is not the official method, but tested it repeatedly, and in this case it does work!)
>

The general instructions are geared towards any situation, including changing wifi_up_down or upstream wifi connections.
> iiab-diagnostics: http://sprunge.us/r0BpIy?en
> 

Noted jupyterhub error open an issue or I will.
> FWIW the output of `journalctl -b 0 -u networkd-dispatcher` below Line 1408 looks potentially abnormal ? (Or maybe this is normal!)

Looks to be recorded during the install, with iiab-diagnostics run pre-reboot, think it's fine but keep an eye on it.
Now the $64 question does the hotspot work?

**Comment by holta** [comment]
> > iiab-diagnostics: http://sprunge.us/r0BpIy?en
> 
> Noted jupyterhub error open an issue or I will.

I missed that.  Thanks for noticing.

**Comment by holta** [comment]
FYI @EMG70 also tested changing the channel using Admin Console — it worked every time — using these instructions:

- Browse to: http://box.lan/admin
  - (Log in to Admin Console with any username that has *sudo* access, i.e. root privileges.)
- Configure (top-left)
- Internal Wi-Fi AP (left side)
- Internal Wi-Fi Access Point Channel (CHANGE IT !)
- Save Configuration (green button)
- Install Configured Options (red button)

<br>

- Utilities (top-right)
- Display Job Status (left side)
- Refresh Status (blue button, until it's done, in about 3-4 min)

<br>

- Verify new channel appears in `/etc/hostapd/hostapd.conf` ***(confirmed!)***

iiab-diagnostics: http://sprunge.us/k2PAnA?en

Finally he also tried connecting a WiFi client devices to this IIAB's hotspot, and it worked, including browsing to the IIAB home page 🎆

**Comment by holta** [comment]
Any other OS's or scenarios, that are important to test ❔

**Comment by jvonau** [comment]
> Any other OS's or scenarios, that are important to test grey_question

I'd merge for wider testing, if you want to cover all based test on RasPiOS also. What would be interesting is RasPi desktop with the NetworkManager back-end enabled, I tested that option when it became available but don't have the time to circle back.

**Comment by holta** [comment]
> test on RasPiOS also

@EMG70 would you have time to test likewise on 64-bit RasPiOS (either "Lite" or "with desktop") in coming days?

**Comment by jvonau** [comment]
> > > iiab-diagnostics: http://sprunge.us/r0BpIy?en
> > 
> > 
> > Noted jupyterhub error open an issue or I will.
> 
> I missed that. Thanks for noticing.

I have a fix available to test

**Comment by holta** [comment]
1) @EMG70 also tested channel-changing using Admin Console on Debian 11 (with desktop) and it now works great ✅ 

2) He will try to test 64-bit RasPiOS with desktop if he has time today or in coming days.

**Comment by holta** [comment]
@jvonau can you take a look at `iiab-test-wifi` output in the 2 screenshots — that @EMG70 captured from the 2 above PR3549-testing machines?  (iiab-diagnostics for each is above)

Debian 11 laptop:
IMG-20230426-WA0005~2

Ubuntu Desktop 23.04 on RPi 400:
IMG-20230426-WA0006~2

(Does verbiage need clarification — e.g. if iiab-test-wifi output is possibly... not reliable in some cases?)

**Comment by holta** [comment]
Fresh install on Ubuntu Server 23.04 VM failed:

> TASK [network : Set Wifi Region country code for hostapd when present] *********
fatal: [127.0.0.1]: FAILED! => {"msg": "The conditional check 'REG_DOM is defined and REG_DOM.stdout | length > 0' failed. The error was: error while evaluating conditional (REG_DOM is defined and REG_DOM.stdout | length > 0): 'dict object' has no attribute 'stdout'.
'dict object' has no attribute 'stdout'\n\nThe error appears to be in '/opt/iiab/iiab/roles/network/tasks/sysd-netd-debian.yml': line 96, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set Wifi Region country code for hostapd when present\n  ^ heren"}

iiab-diagnostics: http://sprunge.us/Ve5UBz?en

**Comment by jvonau** [comment]
> @jvonau can you take a look at `iiab-test-wifi` output in the 2 screenshots — that @EMG70 captured from the 2 above PR3549-testing machines? (iiab-diagnostics for each is above)

> (Does verbiage need clarification — e.g. if iiab-test-wifi output is possibly... not reliable in some cases?)

Need some context, the above would be normal when upstream WiFi is not active, there would be no 'client ssid' to detect.

**Comment by jvonau** [comment]
> Fresh install on Ubuntu Server 23.04 VM failed:
> 
> > TASK [network : Set Wifi Region country code for hostapd when present] *********
> > fatal: [127.0.0.1]: FAILED! => {"msg": "The conditional check 'REG_DOM is defined and REG_DOM.stdout | length > 0' failed. The error was: error while evaluating conditional (REG_DOM is defined and REG_DOM.stdout | length > 0): 'dict object' has no attribute 'stdout'.
> > 'dict object' has no attribute 'stdout'\n\nThe error appears to be in '/opt/iiab/iiab/roles/network/tasks/sysd-netd-debian.yml': line 96, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set Wifi Region country code for hostapd when present\n  ^ heren"}
> 
> iiab-diagnostics: http://sprunge.us/Ve5UBz?en

Hum, don't think 3549 was applied, the ordering is not the same as in this PR.

**Comment by jvonau** [comment]
Just stop making novice testers jump through hoops, 348974be6fab3b0a11b986a3da98f7c3437a48a4 would be the fix for the above error in this PR

**Comment by jvonau** [comment]
>  14 /etc/iiab/pr-list-pulled:
  15 2023-04-26 09:08:40 EDT #3554 4 0f8fdb72 https://github.com/jvonau/iiab jupyterhub "Jupyterhub"

This test was just using 3554, any fixes in this PR would not part of the test environment. Should of used both PRs with curl or iiab. On a positive note 3554 worked correctly to install jupyterhub

**Comment by holta** [comment]
> Should of used both PRs with curl or iiab.

_Right.  Entirely my mistake._

Indeed applying both PRs (#3549 and #3554) worked!

FYI @EMG70 is learning fast, and is a huge help.

> On a positive note 3554 worked correctly to install jupyterhub

💯

**Comment by jvonau** [comment]
Please push both PRs for wider testing, 3549 has the bug fix for the noted error above.

**Comment by holta** [comment]
> Please push both PRs for wider testing, 3549 has the bug fix for the noted error above.

@EMG70 is working on testing 64-bit RasPiOS "with desktop" as a failsafe.

Hopefully both PR's can then be merged immediately after that.

**Comment by holta** [comment]
_Good News:_ Admin Console (and equivalently, `sudo iiab-network`) can sometimes change the Raspberry Pi's WiFi channel.  In this case on @EMG70's 64-bit Raspberry Pi OS (Desktop version) on RPi 400.

_Bad News:_ The entire process fails midstream (intermittently it appears, depending on certain conditions) with the following error:

> 2023-04-26 23:09:26,150 p=54452 u=root n=ansible | fatal: [127.0.0.1]: FAILED! => {"msg": "The conditional check 'country_code is defined and country_code.stdout | length > 0' failed. The error was: error while evaluating conditional (country_code is defined and country_code.stdout | length > 0): 'dict object' has no attribute 'stdout'. 'dict object' has no attribute 'stdout'\n\nThe error appears to be in '/opt/iiab/iiab/roles/network/tasks/rpi_debian.yml': line 44, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set country code for hostapd to value found in /etc/wpa_supplicant/wpa_supplicant.conf\n  ^ here\n"}

iiab-diagnostics: http://sprunge.us/BoVtpB?en

**Comment by holta** [comment]
@jvonau can you review proposed fix 8eb29a9 ?

CLARIF: I logged in to @EMG70's machine and threw in a couple debug statements to make more clear (below) how `sudo iiab-network` failed consistently under certain conditions:

> TASK [network : debug] **********************************************************************************************************************************************************************************************************************
ok: [127.0.0.1] => {
    "country_code": "VARIABLE IS NOT DEFINED!"
}
>
> TASK [network : New Raspbian requires country code -- check for it] *************************************************************************************************************************************************************************
skipping: [127.0.0.1]
>
> TASK [network : debug] **********************************************************************************************************************************************************************************************************************
ok: [127.0.0.1] => {
    "country_code": {
        "changed": false,
        "skip_reason": "Conditional result was False",
        "skipped": true
    }
}
>
> TASK [network : Set country code for hostapd to value found in /etc/wpa_supplicant/wpa_supplicant.conf] *************************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"msg": "The conditional check 'country_code is defined and country_code.stdout | length > 0' failed. The error was: error while evaluating conditional (country_code is defined and country_code.stdout | length > 0): 'dict object' has no attribute 'stdout'. 'dict object' has no attribute 'stdout'\n\nThe error appears to be in '/opt/iiab/iiab/roles/network/tasks/rpi_debian.yml': line 53, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set country code for hostapd to value found in /etc/wpa_supplicant/wpa_supplicant.conf\n  ^ here\n"}

**Comment by holta** [comment]
> > @jvonau can you take a look at `iiab-test-wifi` output in the 2 screenshots — that @EMG70 captured from the 2 above PR3549-testing machines? (iiab-diagnostics for each is above)
> 
> > (Does verbiage need clarification — e.g. if iiab-test-wifi output is possibly... not reliable in some cases?)
> 
> Need some context, the above would be normal when upstream WiFi is not active, there would be no 'client ssid' to detect.

Can it be renamed to (something like) `iiab-test-upstream-ssid` to make this far less confusing?

And/or what words would be best to add to the confusing output below — making clear what WiFi is being tested for what purpose(s) ?

```
root@box:~# iiab-test-wifi
iiab-test-wifi called
running pid 76992
RasPiOS
ssid is
Couldn't find ssid  to use exiting
```

iiab-diagnostics looks good otherwise: http://sprunge.us/OhCTcg?en (internal hotspot channel-changing now appears to work well!)

**Comment by jvonau** [comment]
> @jvonau can you review proposed fix [8eb29a9](https://github.com/iiab/iiab/commit/8eb29a9edea2b99833473c786c23bd660472d764) ?
> 
> CLARIF: I logged in to @EMG70's machine and threw in a couple debug statements to make more clear (below) how `sudo iiab-network` failed consistently under certain conditions:
> 
> > TASK [network : debug] **********************************************************************************************************************************************************************************************************************
> > ok: [127.0.0.1] => {
> > "country_code": "VARIABLE IS NOT DEFINED!"
> > }
> > TASK [network : New Raspbian requires country code -- check for it] *************************************************************************************************************************************************************************
> > skipping: [127.0.0.1]
> > TASK [network : debug] **********************************************************************************************************************************************************************************************************************
> > ok: [127.0.0.1] => {
> > "country_code": {
> > "changed": false,
> > "skip_reason": "Conditional result was False",
> > "skipped": true
> > }
> > }
> > TASK [network : Set country code for hostapd to value found in /etc/wpa_supplicant/wpa_supplicant.conf] *************************************************************************************************************************************
> > fatal: [127.0.0.1]: FAILED! => {"msg": "The conditional check 'country_code is defined and country_code.stdout | length > 0' failed. The error was: error while evaluating conditional (country_code is defined and country_code.stdout | length > 0): 'dict object' has no attribute 'stdout'. 'dict object' has no attribute 'stdout'\n\nThe error appears to be in '/opt/iiab/iiab/roles/network/tasks/rpi_debian.yml': line 53, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set country code for hostapd to value found in /etc/wpa_supplicant/wpa_supplicant.conf\n  ^ here\n"}

Looks fine, similar issue to 348974be6fab3b0a11b986a3da98f7c3437a48a4 when the check became conditional.

**Comment by jvonau** [comment]
> Can it be renamed to (something like) `iiab-test-upstream-ssid` to make this far less confusing?
> 

I'd rather not rename the file to something else as scanning is not the only function the routine provides.

**Comment by holta** [comment]
df08ecf is an attempt at explaining.

I hope that helps @EMG70 and others avoid confusion.

Is that sufficient?  (Finally if the explanation can be further improved, now or in future, even better!)

**Comment by jvonau** [comment]
I think that wording change would only apply when running the file by hand, when wifi_up_down is False the service is disabled and not run during the boot process.

**Comment by holta** [comment]
> I think that wording change would only apply when running the file by hand, when wifi_up_down is False the service is disabled and not run during the boot process.

Definitely.

Good enough for now?

**Comment by holta** [comment]
I propose to merge this PR in coming hours.

If there are no further recommendations / suggestions / objections ?

**Pull request merged**